packages/@aws-cdk/aws-ec2/lib/cfn-init.ts

```diff
@@ -0,0 +1,387 @@
+  }
+
+  /**
+   * Run a command from an argv array
+   *
+   * You do not need to escape space characters or enclose command parameters in quotes.
+   */
+  public static argvCommand(argv: string[], options: InitCommandOptions = {}): InitCommand {
+  }
+
+  protected constructor() {
+    super();
+  }
+}
+
+/**
+ *
+ * The content can be either inline in the template or the content can be
+ * pulled from a URL.
+ */
+export class InitFileOptions {
+  /**
+   * The name of the owning group for this file.
+   *
+   * Not supported for Windows systems.
+   *
+   * @default 'root'
+   */
+  readonly group?: string;
+
+  /**
+   * The name of the owning user for this file.
+   *
+   * Not supported for Windows systems.
+   *
+   * @default 'root'
+   */
+  readonly owner?: string;
+
+  /**
+   * A six-digit octal value representing the mode for this file.
+   *
+   * Use the first three digits for symlinks and the last three digits for
+   * setting permissions. To create a symlink, specify 120xxx, where xxx
+   * defines the permissions of the target file. To specify permissions for a
+   * file, use the last three digits, such as 000644.
+   *
+   * Not supported for Windows systems.
+   *
+   * @default '000644'
+   */
+  readonly mode?: string;
+}
+
+/**
+ * Create files on the EC2 instance.
+ */
+export class InitFile extends InitElement {
+  /**
+   * Use a literal string as the file content
+   */
+  public static fromString(fileName: string, s: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Write a symlink with the given symlink target
+   */
+  public static symlink(fileName: string, target: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Use a base64 string as the file content
+   */
+  public static fromBase64String(fileName: string, base64: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Use a JSON object as the file content, write it to a JSON file.
+   *
+   * May contain tokens.
+   */
+  public static fromJsonObject(fileName: string, obj: Record<string, any>, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Read a file from disk and use its contents
+   *
+   * The file will be embedded in the template, so care should be taken to not
+   * exceed the template size.
+   *
+   * The file will be base64-encoded if it contains nonprintable characters.
+   */
+  public static fromFileInline(targetFileName: string, sourceFileName: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Create an asset from the given file and use that
+   *
+   * This is appropriate for files that are too large to embed into the template.
+   */
+  public static fromFileAsset(targetFileName: string, sourceFileName: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Download from a URL at instance startup time
+   */
+  public static fromUrl(targetFileName: string, url: string, options: InitFileOptions = {}): InitFile {
```

Comment:
How would this be extended to support auth? A new fromUrlWithAuth, or creating a InitFileUrlOptions that extends InitFileOptions and includes the auth params?

Reply:
I was thinking just a new factory.